File: 每日一题/5.2/838.推多米诺.py
#
# @lc app=leetcode.cn id=838 lang=python3
#
# [838] 推多米诺
#

# @lc code=start
class Solution:
    def pushDominoes(self, d: str) -> str:
        d = list('L' + d + 'R')
        n = len(d)
        pre = 0
        for i in range(1, n):
            if d[i] == '.':
                continue
            if d[i] == d[pre]:
                d[pre:i] = d[i] * (i - pre)
            else:
                if d[i] == 'L':
                    m = (pre + i - 1) // 2
                    d[pre:m + 1] = 'R' * (m - pre + 1)
                    m = (pre + i) // 2 + 1
                    d[m:i] = 'L' * (i - m)
            pre = i
        return ''.join(d[1:n - 1])
# @lc code=end

File: 每日一题/5.2/838.推多米诺.cpp
/*
 * @lc app=leetcode.cn id=838 lang=cpp
 *
 * [838] 推多米诺
 */
#include <bits/stdc++.h>
#include <ranges>
using namespace std;
// @lc code=start
class Solution {
public:
    string pushDominoes(string d) {
        d = 'L' + d + 'R';
        int pre = 0;
        for (int i = 1; i < d.size(); i ++ )
        {
            if (d[i] == '.') continue;
            if (d[i] == d[pre])
            {
                fill(d.begin() + pre + 1, d.begin() + i, d[i]);
            }
            else
            {
                if (d[i] == 'L')
                {
                    fill(d.begin() + pre + 1, d.begin() + (pre + i + 1) / 2, 'R');
                    fill(d.begin() + (i + pre) / 2 + 1, d.begin() + i, 'L');
                }
            }
            pre = i;
        }
        return d.substr(1, d.size() - 2);
    }
};
// @lc code=end

